counterparty-core/counterpartycore/lib/transaction_helper/p2sh_serializer.py

```diff
@@ -301,3 +314,295 @@ def make_standard_p2sh_multisig_script(multisig_pubkeys, multisig_pubkeys_requir
+                raise exceptions.TransactionError(
+                    "Unable to select source input for p2sh source address"
+                )
+
+        unsigned_datatx = serialise_p2sh_datatx(
+            pretx_txid,
+            source=source_address,
+            source_input=source_input,
+            destination_outputs=destination_outputs,
+            data_output=data_output,
+            pubkey=dust_return_pubkey,
+            multisig_pubkeys=p2sh_source_multisig_pubkeys,
+            multisig_pubkeys_required=p2sh_source_multisig_pubkeys_required,
+        )
+        unsigned_datatx_hex = binascii.hexlify(unsigned_datatx).decode("utf-8")
+
+        # let the rest of the code work it's magic on the data tx
+        unsigned_tx_hex = unsigned_datatx_hex
+        return pretx_txid, unsigned_pretx_hex, unsigned_tx_hex
+    else:
+        # we're just gonna return the pretx, it doesn't require any of the further checks
+        return pretx_txid, unsigned_pretx_hex, None
+
+
+def serialise_p2sh_datatx(
+    txid,
+    source,
+    source_input,
+    destination_outputs,
+    data_output,
+    pubkey=None,
+    multisig_pubkeys=None,
+    multisig_pubkeys_required=None,
+):
+    assert data_output  # we don't do this unless there's data
+
+    txhash = bitcoinlib.core.lx(bitcoinlib.core.b2x(txid))  # reverse txId
+    data_array, value = data_output
+
+    # version
+    s = (1).to_bytes(4, byteorder="little")
+
+    # number of inputs is the length of data_array (+1 if a source_input exists)
+    number_of_inputs = len(data_array)
+    if source_input is not None:
+        number_of_inputs += 1
+    s += var_int(number_of_inputs)
+
+    # Handle a source input here for a P2SH source
+    if source_input is not None:
+        s += binascii.unhexlify(bytes(source_input["txid"], "utf-8"))[::-1]  # TxOutHash
+        s += source_input["vout"].to_bytes(4, byteorder="little")  # TxOutIndex
+
+        # since pubkey is not returned from indexd, add it from bitcoind
+        source_inputs = script.ensure_script_pub_key_for_inputs([source_input])
+        source_input = source_inputs[0]
+        tx_script = binascii.unhexlify(bytes(source_input["script_pub_key"], "utf-8"))
+        s += var_int(int(len(tx_script)))  # Script length
+        s += tx_script  # Script
+        s += b"\xff" * 4  # Sequence
+
+    # list of inputs
+    for n, data_chunk in enumerate(data_array):
+        data_chunk = config.PREFIX + data_chunk  # prefix the data_chunk  # noqa: PLW2901
+
+        # get the scripts
+        script_sig, redeem_script, output_script = make_p2sh_encoding_redeemscript(
+            data_chunk, n, pubkey, multisig_pubkeys, multisig_pubkeys_required
+        )
+        # substituteScript = script_sig + output_script
+
+        s += txhash  # TxOutHash
+        s += (n).to_bytes(4, byteorder="little")  # TxOutIndex (assumes 0-based)
+
+        # s += var_int(len(substituteScript))                      # Script length
```

Comment:
I think this commented code in this area causes confusion. If you need for testing/debugging can we find a way to make that apparent?

Reply:
I really don't know this code! I just changed the filename.. p2sh transaction still a kind of black box for me..
but agreed it's confusing.. I deleted commented code without explanation..